commit/merkleroot/rmn/client.go

```diff
@@ -19,13 +33,533 @@ type Client interface {
+			// Timer expired, send the observation requests to the rest of the RMN nodes.
+			requestsPerNode := make(map[uint32][]*rmnpb.FixedDestLaneUpdateRequest)
+			for sourceChain, lur := range lursPerChain {
+				for nodeID := range lur.rmnNodes.Iter() {
+					if requestedNodes[sourceChain].Contains(nodeID) {
+						continue
+					}
+					requestedNodes[sourceChain].Add(nodeID)
+					if _, ok := requestsPerNode[nodeID]; !ok {
+						requestsPerNode[nodeID] = make([]*rmnpb.FixedDestLaneUpdateRequest, 0)
+					}
+					requestsPerNode[nodeID] = append(requestsPerNode[nodeID], lur.lur)
+				}
+			}
+			newRequestIDs := c.sendObservationRequests(destChain, requestsPerNode)
+			requestIDs = requestIDs.Union(newRequestIDs)
+		case <-ctx.Done():
+			return nil, ErrTimeout
+		}
+	}
+}
+
+func (c *PBClient) validateSignedObservationResponse(
+	rmnNodeID uint32,
+	lurs map[uint64]lurWithMeta,
+	signedObs *rmnpb.SignedObservation,
+	destChain *rmnpb.LaneDest,
+) error {
+	if signedObs.Observation.LaneDest != destChain {
+		return fmt.Errorf("unexpected lane dest %v", signedObs.Observation.LaneDest)
+	}
+
+	if !bytes.Equal(signedObs.Observation.RmnHomeContractConfigDigest, c.rmnHomeConfigDigest) {
+		return fmt.Errorf("unexpected rmn home contract config digest %v",
+			signedObs.Observation.RmnHomeContractConfigDigest)
+	}
+
+	for _, signedObsLu := range signedObs.Observation.FixedDestLaneUpdates {
+		lur, exists := lurs[signedObsLu.LaneSource.SourceChainSelector]
+		if !exists {
+			return fmt.Errorf("unexpected source chain selector %d", signedObsLu.LaneSource.SourceChainSelector)
+		}
+
+		if !lur.rmnNodes.Contains(rmnNodeID) {
+			return fmt.Errorf("rmn node %d not expected to read chain %d",
+				rmnNodeID, signedObsLu.LaneSource.SourceChainSelector)
+		}
+
+		if lur.lur.ClosedInterval != signedObsLu.ClosedInterval {
+			return fmt.Errorf("unexpected closed interval %v", signedObsLu.ClosedInterval)
+		}
+		if lur.lur.LaneSource != signedObsLu.LaneSource {
+			return fmt.Errorf("unexpected lane source %v", signedObsLu.LaneSource)
+		}
+		if signedObsLu.Root == nil {
+			return errors.New("root is nil")
+		}
+	}
+	return nil
+}
+
+func (c *PBClient) getRmnReportSignatures(
+	ctx context.Context,
+	destChain *rmnpb.LaneDest,
+	rmnSignedObservations []rmnSignedObservationWithMeta,
+) (*ReportSignatures, error) {
+	// At this point we might have multiple signedObservations for different nodes but never for the same source chain
+	// from the same node.
+	//
+	// e.g.
+	// The following nodes support the following chains and minObservers=2:
+	// node1: [1]	node2:[1,2,3]	node3:[1,2,3]
+	//
+	// node1: getObservations(1)          ->             never_responds
+	// node2: getObservations(1,2,3)      -> responds_ok
+	// node3: getObservations(2,3)        ->     responds_ok
+	// node3: getObservations(1)          ->                         responds_ok (after timeout we made a new request)
+
+	sigObservations := make([]*rmnpb.AttributedSignedObservation, 0)
```

Comment:
style nit: nil is a valid slice (see https://github.com/uber-go/guide/blob/master/style.md#nil-is-a-valid-slice)

Reply:
Didn't realize that was in the uber guide recommended it this way
```suggestion
	var sigObservations []*rmnpb.AttributedSignedObservation
```